File: modules/@angular/core/src/reflection/reflection_capabilities.dart
library reflection.reflection_capabilities;

import 'dart:mirrors';

import 'package:angular2/src/core/metadata/lifecycle_hooks.dart';
import 'package:angular2/src/facade/lang.dart';

import 'platform_reflection_capabilities.dart';
import 'types.dart';

import '../linker/template_ref.dart';

var DOT_REGEX = new RegExp('\\.');

class ReflectionCapabilities implements PlatformReflectionCapabilities {
  Map<Symbol, Type> parameterizedTypeMapping = new Map<Symbol, Type>();

  ReflectionCapabilities([metadataReader]) {
    // In Dart, there is no way of getting from a parameterized Type to
    // the underlying non parameterized type.
    // So we need to have a separate Map for the types that are generic
    // and used in our DI...
    parameterizedTypeMapping[reflectType(TemplateRef).qualifiedName] = TemplateRef;
  }

  _typeFromMirror(TypeMirror typeMirror) {
    var result = parameterizedTypeMapping[typeMirror.qualifiedName];
    if (result == null && typeMirror.hasReflectedType && typeMirror.reflectedType != dynamic) {
      result = typeMirror.reflectedType;
    }
    return result;
  }

  bool isReflectionEnabled() {
    return true;
  }

  Function factory(Type type) {
    ClassMirror classMirror = reflectType(type);
    MethodMirror ctor = classMirror.declarations[classMirror.simpleName];
    Function create = classMirror.newInstance;
    Symbol name = ctor.constructorName;
    int length = ctor.parameters.length;

    switch (length) {
      case 0:
        return () => create(name, []).reflectee;
      case 1:
        return (a1) => create(name, [a1]).reflectee;
      case 2:
        return (a1, a2) => create(name, [a1, a2]).reflectee;
      case 3:
        return (a1, a2, a3) => create(name, [a1, a2, a3]).reflectee;
      case 4:
        return (a1, a2, a3, a4) => create(name, [a1, a2, a3, a4]).reflectee;
      case 5:
        return (a1, a2, a3, a4, a5) =>
            create(name, [a1, a2, a3, a4, a5]).reflectee;
      case 6:
        return (a1, a2, a3, a4, a5, a6) =>
            create(name, [a1, a2, a3, a4, a5, a6]).reflectee;
      case 7:
        return (a1, a2, a3, a4, a5, a6, a7) =>
            create(name, [a1, a2, a3, a4, a5, a6, a7]).reflectee;
      case 8:
        return (a1, a2, a3, a4, a5, a6, a7, a8) =>
            create(name, [a1, a2, a3, a4, a5, a6, a7, a8]).reflectee;
      case 9:
        return (a1, a2, a3, a4, a5, a6, a7, a8, a9) =>
            create(name, [a1, a2, a3, a4, a5, a6, a7, a8, a9]).reflectee;
      case 10:
        return (a1, a2, a3, a4, a5, a6, a7, a8, a9, a10) =>
            create(name, [a1, a2, a3, a4, a5, a6, a7, a8, a9, a10]).reflectee;
      case 11:
        return (a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11) =>
            create(name, [a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11])
                .reflectee;
      case 12:
        return (a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12) =>
            create(name, [a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12])
                .reflectee;
      case 13:
        return (a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13) =>
            create(name, [
              a1,
              a2,
              a3,
              a4,
              a5,
              a6,
              a7,
              a8,
              a9,
              a10,
              a11,
              a12,
              a13
            ]).reflectee;
      case 14:
        return (a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14) =>
            create(name, [
              a1,
              a2,
              a3,
              a4,
              a5,
              a6,
              a7,
              a8,
              a9,
              a10,
              a11,
              a12,
              a13,
              a14
            ]).reflectee;
      case 15:
        return (a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14,
                a15) =>
            create(name, [
              a1,
              a2,
              a3,
              a4,
              a5,
              a6,
              a7,
              a8,
              a9,
              a10,
              a11,
              a12,
              a13,
              a14,
              a15
            ]).reflectee;
      case 16:
        return (a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14,
                a15, a16) =>
            create(name, [
              a1,
              a2,
              a3,
              a4,
              a5,
              a6,
              a7,
              a8,
              a9,
              a10,
              a11,
              a12,
              a13,
              a14,
              a15,
              a16
            ]).reflectee;
      case 17:
        return (a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14,
                a15, a16, a17) =>
            create(name, [
              a1,
              a2,
              a3,
              a4,
              a5,
              a6,
              a7,
              a8,
              a9,
              a10,
              a11,
              a12,
              a13,
              a14,
              a15,
              a16,
              a17
            ]).reflectee;
      case 18:
        return (a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14,
                a15, a16, a17, a18) =>
            create(name, [
              a1,
              a2,
              a3,
              a4,
              a5,
              a6,
              a7,
              a8,
              a9,
              a10,
              a11,
              a12,
              a13,
              a14,
              a15,
              a16,
              a17,
              a18
            ]).reflectee;
      case 19:
        return (a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14,
                a15, a16, a17, a18, a19) =>
            create(name, [
              a1,
              a2,
              a3,
              a4,
              a5,
              a6,
              a7,
              a8,
              a9,
              a10,
              a11,
              a12,
              a13,
              a14,
              a15,
              a16,
              a17,
              a18,
              a19
            ]).reflectee;
      case 20:
        return (a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14,
                a15, a16, a17, a18, a19, a20) =>
            create(name, [
              a1,
              a2,
              a3,
              a4,
              a5,
              a6,
              a7,
              a8,
              a9,
              a10,
              a11,
              a12,
              a13,
              a14,
              a15,
              a16,
              a17,
              a18,
              a19,
              a20
            ]).reflectee;
    }

    throw "Cannot create a factory for '${stringify(type)}' because its constructor has more than 20 arguments";
  }

  List<List> parameters(typeOrFunc) {
    final parameters = typeOrFunc is Type
        ? _constructorParameters(typeOrFunc)
        : _functionParameters(typeOrFunc);
    return parameters.map(_convertParameter).toList();
  }

  List _convertParameter(ParameterMirror p) {
    var t = p.type;
    var type = _typeFromMirror(t);
    var res = type != null ? [type] : [];
    res.addAll(p.metadata.map((m) => m.reflectee));
    return res;
  }

  List annotations(typeOrFunc) {
    final meta = typeOrFunc is Type
        ? _constructorMetadata(typeOrFunc)
        : _functionMetadata(typeOrFunc);

    return meta.map((m) => m.reflectee).toList();
  }

  Map propMetadata(typeOrFunc) {
    final res = {};
    reflectClass(typeOrFunc).declarations.forEach((k, v) {
      var name = _normalizeName(MirrorSystem.getName(k));
      if (res[name] == null) res[name] = [];
      res[name].addAll(v.metadata.map((fm) => fm.reflectee));
    });
    return res;
  }

  String _normalizeName(String name) {
    return name.endsWith("=") ? name.substring(0, name.length - 1) : name;
  }

  bool hasLifecycleHook(dynamic type, Type lcInterface, String lcProperty) {
    if (type is! Type) return false;
    return this.interfaces(type).contains(lcInterface);
  }

  List interfaces(type) {
    final clazz = reflectType(type);
    _assertDeclaresLifecycleHooks(clazz);
    return _interfacesFromMirror(clazz);
  }

  List _interfacesFromMirror(classMirror) {
    return classMirror.superinterfaces.map((si) => si.reflectedType).toList()
      ..addAll(classMirror.superclass == null
          ? []
          : _interfacesFromMirror(classMirror.superclass));
  }

  GetterFn getter(String name) {
    var symbol = new Symbol(name);
    return (receiver) => reflect(receiver).getField(symbol).reflectee;
  }

  SetterFn setter(String name) {
    var symbol = new Symbol(name);
    return (receiver, value) =>
        reflect(receiver).setField(symbol, value).reflectee;
  }

  MethodFn method(String name) {
    var symbol = new Symbol(name);
    return (receiver, posArgs) =>
        reflect(receiver).invoke(symbol, posArgs).reflectee;
  }

  List _functionParameters(Function func) {
    var closureMirror = reflect(func);
    return closureMirror.function.parameters;
  }

  List _constructorParameters(Type type) {
    ClassMirror classMirror = reflectType(type);
    MethodMirror ctor = classMirror.declarations[classMirror.simpleName];
    return ctor.parameters;
  }

  List _functionMetadata(Function func) {
    var closureMirror = reflect(func);
    return closureMirror.function.metadata;
  }

  List _constructorMetadata(Type type) {
    ClassMirror classMirror = reflectType(type);
    return classMirror.metadata;
  }

  String importUri(dynamic type) {
    // StaticSymbol
    if (type is Map && type['filePath'] != null) {
      return type['filePath'];
    }
    // Runtime type
    return '${(reflectClass(type).owner as LibraryMirror).uri}';
  }
}

final _lifecycleHookMirrors = <ClassMirror>[
  reflectType(AfterContentChecked),
  reflectType(AfterContentInit),
  reflectType(AfterViewChecked),
  reflectType(AfterViewInit),
  reflectType(DoCheck),
  reflectType(OnChanges),
  reflectType(OnDestroy),
  reflectType(OnInit),
];

/// Checks whether [clazz] implements lifecycle ifaces without declaring them.
///
/// Due to Dart implementation details, lifecycle hooks are only called when a
/// class explicitly declares that it implements the associated interface.
/// See https://goo.gl/b07Kii for details.
void _assertDeclaresLifecycleHooks(ClassMirror clazz) {
  final missingDeclarations = <ClassMirror>[];
  for (var iface in _lifecycleHookMirrors) {
    if (!_checkDeclares(clazz, iface: iface) &&
        _checkImplements(clazz, iface: iface)) {
      missingDeclarations.add(iface);
    }
  }
  if (missingDeclarations.isNotEmpty) {
    throw new MissingInterfaceError(clazz, missingDeclarations);
  }
}

/// Returns whether [clazz] declares that it implements [iface].
///
/// Returns `false` if [clazz] implements [iface] but does not declare it.
/// Returns `false` if [clazz]'s superclass declares that it
/// implements [iface].
bool _checkDeclares(ClassMirror clazz, {ClassMirror iface: null}) {
  if (iface == null) {
    throw new ArgumentError.notNull('iface');
  }
  return clazz.superinterfaces.contains(iface);
}

/// Returns whether [clazz] implements [iface].
///
/// Returns `true` if [clazz] implements [iface] and does not declare it.
/// Returns `true` if [clazz]'s superclass implements [iface].
///
/// This is an approximation of a JavaScript feature check:
/// ```js
/// var matches = true;
/// for (var prop in iface) {
///   if (iface.hasOwnProperty(prop)) {
///     matches = matches && clazz.hasOwnProperty(prop);
///   }
/// }
/// return matches;
/// ```
bool _checkImplements(ClassMirror clazz, {ClassMirror iface: null}) {
  if (iface == null) {
    throw new ArgumentError.notNull('iface');
  }

  var matches = true;
  iface.declarations.forEach((symbol, declarationMirror) {
    if (!matches) return;
    if (declarationMirror.isConstructor || declarationMirror.isPrivate) return;
    matches = clazz.declarations.keys.contains(symbol);
  });
  if (!matches && clazz.superclass != null) {
    matches = _checkImplements(clazz.superclass, iface: iface);
  }
  if (!matches && clazz.mixin != clazz) {
    matches = _checkImplements(clazz.mixin, iface: iface);
  }

  return matches;
}

/// Error thrown when a class implements a lifecycle iface it does not declare.
class MissingInterfaceError extends Error {
  final ClassMirror clazz;
  final List<ClassMirror> missingDeclarations;

  MissingInterfaceError(this.clazz, this.missingDeclarations);

  @override
  String toString() {
    final buf = new StringBuffer();
    buf.write('${clazz.simpleName} implements ');
    if (missingDeclarations.length == 1) {
      buf.write('an interface but does not declare it: ');
    } else {
      buf.write('interfaces but does not declare them: ');
    }
    buf.write(
        missingDeclarations.map((d) => d.simpleName.toString()).join(', '));
    buf.write('. See https://goo.gl/b07Kii for more info.');
    return buf.toString();
  }
}

File: modules/@angular/core/src/reflection/reflection_capabilities.ts
import {BaseException} from '../facade/exceptions';
import {ConcreteType, Type, global, isFunction, isPresent, stringify} from '../facade/lang';

import {PlatformReflectionCapabilities} from './platform_reflection_capabilities';
import {GetterFn, MethodFn, SetterFn} from './types';

export class ReflectionCapabilities implements PlatformReflectionCapabilities {
  private _reflect: any;

  constructor(reflect?: any) { this._reflect = isPresent(reflect) ? reflect : global.Reflect; }

  isReflectionEnabled(): boolean { return true; }

  factory(t: ConcreteType): Function {
    switch (t.length) {
      case 0:
        return () => new t();
      case 1:
        return (a1: any) => new t(a1);
      case 2:
        return (a1: any, a2: any) => new t(a1, a2);
      case 3:
        return (a1: any, a2: any, a3: any) => new t(a1, a2, a3);
      case 4:
        return (a1: any, a2: any, a3: any, a4: any) => new t(a1, a2, a3, a4);
      case 5:
        return (a1: any, a2: any, a3: any, a4: any, a5: any) => new t(a1, a2, a3, a4, a5);
      case 6:
        return (a1: any, a2: any, a3: any, a4: any, a5: any, a6: any) =>
                   new t(a1, a2, a3, a4, a5, a6);
      case 7:
        return (a1: any, a2: any, a3: any, a4: any, a5: any, a6: any, a7: any) =>
                   new t(a1, a2, a3, a4, a5, a6, a7);
      case 8:
        return (a1: any, a2: any, a3: any, a4: any, a5: any, a6: any, a7: any, a8: any) =>
                   new t(a1, a2, a3, a4, a5, a6, a7, a8);
      case 9:
        return (a1: any, a2: any, a3: any, a4: any, a5: any, a6: any, a7: any, a8: any, a9: any) =>
                   new t(a1, a2, a3, a4, a5, a6, a7, a8, a9);
      case 10:
        return (a1: any, a2: any, a3: any, a4: any, a5: any, a6: any, a7: any, a8: any, a9: any,
                a10: any) => new t(a1, a2, a3, a4, a5, a6, a7, a8, a9, a10);
      case 11:
        return (a1: any, a2: any, a3: any, a4: any, a5: any, a6: any, a7: any, a8: any, a9: any,
                a10: any, a11: any) => new t(a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11);
      case 12:
        return (a1: any, a2: any, a3: any, a4: any, a5: any, a6: any, a7: any, a8: any, a9: any,
                a10: any, a11: any, a12: any) =>
                   new t(a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12);
      case 13:
        return (a1: any, a2: any, a3: any, a4: any, a5: any, a6: any, a7: any, a8: any, a9: any,
                a10: any, a11: any, a12: any, a13: any) =>
                   new t(a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13);
      case 14:
        return (a1: any, a2: any, a3: any, a4: any, a5: any, a6: any, a7: any, a8: any, a9: any,
                a10: any, a11: any, a12: any, a13: any, a14: any) =>
                   new t(a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14);
      case 15:
        return (a1: any, a2: any, a3: any, a4: any, a5: any, a6: any, a7: any, a8: any, a9: any,
                a10: any, a11: any, a12: any, a13: any, a14: any, a15: any) =>
                   new t(a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14, a15);
      case 16:
        return (a1: any, a2: any, a3: any, a4: any, a5: any, a6: any, a7: any, a8: any, a9: any,
                a10: any, a11: any, a12: any, a13: any, a14: any, a15: any, a16: any) =>
                   new t(a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14, a15, a16);
      case 17:
        return (a1: any, a2: any, a3: any, a4: any, a5: any, a6: any, a7: any, a8: any, a9: any,
                a10: any, a11: any, a12: any, a13: any, a14: any, a15: any, a16: any, a17: any) =>
                   new t(
                       a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14, a15, a16, a17);
      case 18:
        return (a1: any, a2: any, a3: any, a4: any, a5: any, a6: any, a7: any, a8: any, a9: any,
                a10: any, a11: any, a12: any, a13: any, a14: any, a15: any, a16: any, a17: any,
                a18: any) =>
                   new t(
                       a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14, a15, a16, a17,
                       a18);
      case 19:
        return (a1: any, a2: any, a3: any, a4: any, a5: any, a6: any, a7: any, a8: any, a9: any,
                a10: any, a11: any, a12: any, a13: any, a14: any, a15: any, a16: any, a17: any,
                a18: any, a19: any) =>
                   new t(
                       a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14, a15, a16, a17,
                       a18, a19);
      case 20:
        return (a1: any, a2: any, a3: any, a4: any, a5: any, a6: any, a7: any, a8: any, a9: any,
                a10: any, a11: any, a12: any, a13: any, a14: any, a15: any, a16: any, a17: any,
                a18: any, a19: any, a20: any) =>
                   new t(
                       a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14, a15, a16, a17,
                       a18, a19, a20);
    };

    throw new Error(
        `Cannot create a factory for '${stringify(t)}' because its constructor has more than 20 arguments`);
  }

  /** @internal */
  _zipTypesAndAnnotations(
      paramTypes: any /** TODO #9100 */, paramAnnotations: any /** TODO #9100 */): any[][] {
    var result: any /** TODO #9100 */;

    if (typeof paramTypes === 'undefined') {
      result = new Array(paramAnnotations.length);
    } else {
      result = new Array(paramTypes.length);
    }

    for (var i = 0; i < result.length; i++) {
      // TS outputs Object for parameters without types, while Traceur omits
      // the annotations. For now we preserve the Traceur behavior to aid
      // migration, but this can be revisited.
      if (typeof paramTypes === 'undefined') {
        result[i] = [];
      } else if (paramTypes[i] != Object) {
        result[i] = [paramTypes[i]];
      } else {
        result[i] = [];
      }
      if (isPresent(paramAnnotations) && isPresent(paramAnnotations[i])) {
        result[i] = result[i].concat(paramAnnotations[i]);
      }
    }
    return result;
  }

  parameters(typeOrFunc: Type): any[][] {
    // Prefer the direct API.
    if (isPresent((<any>typeOrFunc).parameters)) {
      return (<any>typeOrFunc).parameters;
    }

    // API of tsickle for lowering decorators to properties on the class.
    if (isPresent((<any>typeOrFunc).ctorParameters)) {
      let ctorParameters = (<any>typeOrFunc).ctorParameters;
      let paramTypes =
          ctorParameters.map((ctorParam: any /** TODO #9100 */) => ctorParam && ctorParam.type);
      let paramAnnotations = ctorParameters.map(
          (ctorParam: any /** TODO #9100 */) =>
              ctorParam && convertTsickleDecoratorIntoMetadata(ctorParam.decorators));
      return this._zipTypesAndAnnotations(paramTypes, paramAnnotations);
    }

    // API for metadata created by invoking the decorators.
    if (isPresent(this._reflect) && isPresent(this._reflect.getMetadata)) {
      var paramAnnotations = this._reflect.getMetadata('parameters', typeOrFunc);
      var paramTypes = this._reflect.getMetadata('design:paramtypes', typeOrFunc);
      if (isPresent(paramTypes) || isPresent(paramAnnotations)) {
        return this._zipTypesAndAnnotations(paramTypes, paramAnnotations);
      }
    }
    // The array has to be filled with `undefined` because holes would be skipped by `some`
    let parameters = new Array((<any>typeOrFunc.length));
    parameters.fill(undefined);
    return parameters;
  }

  annotations(typeOrFunc: Type): any[] {
    // Prefer the direct API.
    if (isPresent((<any>typeOrFunc).annotations)) {
      var annotations = (<any>typeOrFunc).annotations;
      if (isFunction(annotations) && annotations.annotations) {
        annotations = annotations.annotations;
      }
      return annotations;
    }

    // API of tsickle for lowering decorators to properties on the class.
    if (isPresent((<any>typeOrFunc).decorators)) {
      return convertTsickleDecoratorIntoMetadata((<any>typeOrFunc).decorators);
    }

    // API for metadata created by invoking the decorators.
    if (isPresent(this._reflect) && isPresent(this._reflect.getMetadata)) {
      var annotations = this._reflect.getMetadata('annotations', typeOrFunc);
      if (isPresent(annotations)) return annotations;
    }
    return [];
  }

  propMetadata(typeOrFunc: any): {[key: string]: any[]} {
    // Prefer the direct API.
    if (isPresent((<any>typeOrFunc).propMetadata)) {
      var propMetadata = (<any>typeOrFunc).propMetadata;
      if (isFunction(propMetadata) && propMetadata.propMetadata) {
        propMetadata = propMetadata.propMetadata;
      }
      return propMetadata;
    }

    // API of tsickle for lowering decorators to properties on the class.
    if (isPresent((<any>typeOrFunc).propDecorators)) {
      let propDecorators = (<any>typeOrFunc).propDecorators;
      let propMetadata = <{[key: string]: any[]}>{};
      Object.keys(propDecorators).forEach(prop => {
        propMetadata[prop] = convertTsickleDecoratorIntoMetadata(propDecorators[prop]);
      });
      return propMetadata;
    }

    // API for metadata created by invoking the decorators.
    if (isPresent(this._reflect) && isPresent(this._reflect.getMetadata)) {
      var propMetadata = this._reflect.getMetadata('propMetadata', typeOrFunc);
      if (isPresent(propMetadata)) return propMetadata;
    }
    return {};
  }

  // Note: JavaScript does not support to query for interfaces during runtime.
  // However, we can't throw here as the reflector will always call this method
  // when asked for a lifecycle interface as this is what we check in Dart.
  interfaces(type: Type): any[] { return []; }

  hasLifecycleHook(type: any, lcInterface: Type, lcProperty: string): boolean {
    if (!(type instanceof Type)) return false;

    var proto = (<any>type).prototype;
    return !!proto[lcProperty];
  }

  getter(name: string): GetterFn { return <GetterFn>new Function('o', 'return o.' + name + ';'); }

  setter(name: string): SetterFn {
    return <SetterFn>new Function('o', 'v', 'return o.' + name + ' = v;');
  }

  method(name: string): MethodFn {
    let functionBody = `if (!o.${name}) throw new Error('"${name}" is undefined');
        return o.${name}.apply(o, args);`;
    return <MethodFn>new Function('o', 'args', functionBody);
  }

  // There is not a concept of import uri in Js, but this is useful in developing Dart applications.
  importUri(type: any): string {
    // StaticSymbol
    if (typeof type === 'object' && type['filePath']) {
      return type['filePath'];
    }
    // Runtime type
    return `./${stringify(type)}`;
  }
}

function convertTsickleDecoratorIntoMetadata(decoratorInvocations: any[]): any[] {
  if (!decoratorInvocations) {
    return [];
  }
  return decoratorInvocations.map(decoratorInvocation => {
    var decoratorType = decoratorInvocation.type;
    var annotationCls = decoratorType.annotationCls;
    var annotationArgs = decoratorInvocation.args ? decoratorInvocation.args : [];
    var annotation = Object.create(annotationCls.prototype);
    annotationCls.apply(annotation, annotationArgs);
    return annotation;
  });
}
